File: beams/www/job_application_upload/upload_doc.py
import datetime
import json

import frappe
from frappe import _
from frappe.utils import escape_html
from frappe.utils.file_manager import save_file
from frappe.utils.password import decrypt


def get_context(context):
	'''
		Initializes the context with job applicant details if the applicant ID is valid.
		args:
			context (dict)
		Return : None
	'''
	context.no_cache = 1
	context.doc = {}
	encrypted_applicant_id = frappe.form_dict.applicant_id
	applicant_id = authorize_applicant_id(encrypted_applicant_id)
	if frappe.db.exists('Job Applicant', applicant_id):
		context.doc = frappe.get_doc('Job Applicant', applicant_id)
	context.languages = frappe.db.get_all('Language', fields=['language_name', 'name'])

def authorize_applicant_id(encrypted_applicant_id):
	'''
		Method is used to check the magic link exist or expired
		args:
			encrypted_applicant_id: encrypted value of Applicant ID
		return Applicant ID (example: 123)
	'''
	decrypted_applicant_id = False
	try:
		decrypted_applicant_id = decrypt(encrypted_applicant_id)
	except Exception:
		frappe.log_error(str(encrypted_applicant_id), 'Dirty Applicant Link')
		frappe.throw(_('Sorry, we were not able to decrypt the applicant link'), frappe.PermissionError)

	if decrypted_applicant_id:
		if not frappe.db.exists('Job Applicant', decrypted_applicant_id):
			frappe.log_error(str(decrypted_applicant_id), 'Job Applicant Not exist')
			frappe.throw(_('Sorry, couldn\'t find any matching Job Applicant'), frappe.PermissionError)
		else:
			is_form_submitted = frappe.db.get_value('Job Applicant', decrypted_applicant_id, 'is_form_submitted')
			if is_form_submitted:
				frappe.log_error(str(decrypted_applicant_id), 'Form Already Submitted')
				frappe.throw(_('Sorry, Form is already Submitted'), frappe.PermissionError)
			else:
				return decrypted_applicant_id

@frappe.whitelist(allow_guest=True)
def update_register_form(docname, form_data=None):
	'''
		Updates a Job Applicant document with form data, handling various sections and child tables.
	'''
	
	try:
		form_data_str = frappe.form_dict.form_data
		form_data = json.loads(form_data_str)

		docname = form_data.get('docname')
		if not docname:
			return {'status': 'error', 'message': 'No document name provided.'}

		if not form_data.get('applicant_name'):
			frappe.throw('Missing required field: Applicant Name')

		doc = frappe.get_doc('Job Applicant', docname)

		def sanitize(field):
			return escape_html(form_data.get(field) or '')

		def convert_date(date_str):
			return datetime.datetime.strptime(date_str, '%d-%m-%Y').strftime('%Y-%m-%d') if date_str else None

		doc.applicant_name = sanitize('applicant_name')
		doc.father_name = sanitize('father_name')
		doc.date_of_birth = convert_date(form_data.get('date_of_birth'))
		doc.gender = sanitize('gender')
		doc.country = sanitize('country')
		doc.marital_status = sanitize('marital_status')
		doc.aadhar_number = sanitize('aadhaar_number_input')

		doc.house_no_name = sanitize('current_house_no')
		doc.street_road = sanitize('current_street')
		doc.locality_village = sanitize('current_locality')
		doc.city = sanitize('current_city')
		doc.district = sanitize('current_district')
		doc.state = sanitize('current_state')
		doc.post_office = sanitize('current_perm_post_office')
		doc.pin_code = sanitize('current_pin')

		doc.phouse_no_name = sanitize('permanent_house_no')
		doc.pstreet_road = sanitize('permanent_street')
		doc.plocality_village = sanitize('permanent_locality')
		doc.pcity = sanitize('permanent_city')
		doc.pdistrict = sanitize('permanent_district')
		doc.pstate = sanitize('permanent_state')
		doc.ppost_office = sanitize('permanent_perm_post_office')
		doc.ppin_code = sanitize('permanent_pin')

		doc.name_of_employer = sanitize('name_of_employer')
		doc.department = sanitize('current_department')
		doc.cdesignation = sanitize('current_designation')
		doc.reports_to = sanitize('reports_to')
		doc.cname = sanitize('manager_name')
		doc.ccontact = sanitize('manager_contact_no')
		doc.cemail = sanitize('manager_email')
		doc.creference = sanitize('reference_taken')
		doc.address_of_employeer = sanitize('address_of_employer')
		doc.duties_and_reponsibilitiess = sanitize('duties_and_responsibilities')
		doc.reason_for_leavingg = sanitize('reason_for_leaving')
		doc.current_employment_type = sanitize('was_this_position')
		doc.agency_details = sanitize('agency_details')
		doc.phone_number = sanitize('phone_number')
		doc.current_salary = sanitize('current_salary')
		doc.first_salary_drawn = sanitize('first_salary_drawn')
		doc.telephone_number = sanitize('telephone_number')
		doc.email_id = sanitize('email_id')
		doc.employee_code = sanitize('employee_code')
		doc.in_india = bool(form_data.get('in_india'))
		doc.abroad = bool(form_data.get('abroad'))
		doc.is_form_submitted = form_data.get('is_form_submitted')

		doc.expected_salary = sanitize('expected_salary')
		doc.other_achievments = sanitize('other_achievments')
		doc.position = sanitize('position')
		doc.interviewed_location = sanitize('interviewed_location')
		doc.interviewed_date = convert_date(form_data.get('interviewed_date'))
		doc.interviewed_outcome = sanitize('interviewed_outcome')
		doc.related_employee = sanitize('related_employee')
		doc.related_employee_org = sanitize('related_employee_org')
		doc.related_employee_pos = sanitize('related_employee_pos')
		doc.related_employee_rel = sanitize('related_employee_rel')
		doc.professional_org = sanitize('professional_org')
		doc.political_org = sanitize('political_org')
		doc.specialised_training = sanitize('specialised_training')
		doc.share_your_thoughts = sanitize('additional_comments')

		years = form_data.get('period_years') or '0'
		months = form_data.get('current_period_months') or '0'
		doc.period_of_stay = f'{years} Years {months} Months'

		for field in ['payslip_month_1', 'payslip_month_2', 'payslip_month_3']:
			if form_data.get(field):
				filename = update_file(form_data[field], 'Job Applicant', docname) or ''
				setattr(doc, field, filename)

		doc.education_qualification = []
		for row in form_data.get('education_qualification', []):
			if row.get('course'):
				filename = update_file(row.get('attachments'), 'Job Applicant', docname) or ''
				doc.append('education_qualification', {
					'course': row.get('course'),
					'name_of_school_college': row.get('name_of_school_college'),
					'name_of_universityboard_of_exam': row.get('name_of_universityboard_of_exam'),
					'dates_attended_from': int(row.get('dates_attended_from')),
					'dates_attended_to': int(row.get('dates_attended_to')),
					'result': float(row.get('result')),
					'attachments': filename
				})

		doc.professional_certification = []
		for row in form_data.get('professional_certification', []):
			if row.get('course'):
				filename = update_file(row.get('attachments'), 'Job Applicant', docname) or ''
				doc.append('professional_certification', {
					'course': row.get('course'),
					'institute_name': row.get('institute_name'),
					'dates_attended_from': int(row.get('dates_attended_from')),
					'dates_attended_to': int(row.get('dates_attended_to')),
					'type_of_certification': row.get('type_of_certification'),
					'subject_major': row.get('subject_major'),
					'attachments': filename
				})

		doc.prev_emp_his = []
		for row in form_data.get('prev_emp_his', []):
			if row.get('name_of_org'):
				filename = update_file(row.get('attachments'), 'Job Applicant', docname) or ''
				doc.append('prev_emp_his', {
					'name_of_org': row.get('name_of_org'),
					'prev_designation': row.get('prev_designation'),
					'last_salary_drawn': row.get('last_salary_drawn'),
					'name_of_manager': row.get('name_of_manager'),
					'period_of_employment': row.get('period_of_employment'),
					'reason_for_leaving': row.get('reason_for_leaving'),
					'attachments': filename
				})

		doc.language_proficiency = []
		for row in form_data.get('language_proficiency', []):
			doc.append('language_proficiency', {
				'language': row.get('language'),
				'speak': row.get('speak'),
				'read': row.get('read'),
				'write': row.get('write')
			})

		doc.save(ignore_permissions=True)
		frappe.db.commit()

		return {'message': 'success', 'docname': doc.name}

	except Exception as e:
		frappe.log_error(
			title='Job Application Update Failed',
			message=f'Error updating applicant \'{docname}\'\nException: {str(e)}'
		)
		return {'message': str(e)}

@frappe.whitelist(allow_guest=True)
def update_file(filedata, doctype, docname):
	file_name = None
	if filedata and doctype and docname:
		filedata_list = filedata['files_data']
		for filedata_item in filedata_list:
			filedoc = save_file(
				filedata_item['filename'],
				filedata_item['dataurl'],
				doctype,
				docname,
				decode=True,
				is_private=0
			)
			file_name = filedoc.file_url
	frappe.db.commit()
	return file_name

File: beams/www/job_application_upload/upload_doc.js
$(document).ready(function () {
	$('.navbar').hide();
	$('.web-footer').hide();
	const { get_query_params, get_query_string } = frappe.utils;
	const applicant_id = $('#docname').val();

	// Handle file selection and reading for each file input
	var $form = $('form[id="submit_application"]');
	$form.on('change', '[type="file"]', function () {
		var $input = $(this);
		var input = $input.get(0);
		if (input.files.length) {
			input.filedata = { 'files_data': [] };
			$.each(input.files, function (key, value) {
				setupReader(value, input);
			});
		}
	});

	function setupReader(file, input) {
		var reader = new FileReader();
		reader.onload = function (e) {
			input.filedata.files_data.push({
				'__file_attachment': 1,
				'filename': file.name,
				'dataurl': reader.result
			});
		};
		reader.readAsDataURL(file);
	}

	// Safely sanitize values
	const safeValue = (value) => value ? frappe.utils.xss_sanitise(String(value)) : '';
	const date_of_birth = safeValue($('#date_of_birth').val());
	const interviewed_date = safeValue($('#interviewed_date').val());

	$form.on('submit', function (event) {
		event.preventDefault();

		const currentEmployerTab = document.getElementById("currentEmployerTab");
		const aadhaarField = document.getElementById('aadhaar_field');
		const aadhaarInput = document.getElementById('aadhaar_number_input');

		const isAadhaarVisible = aadhaarField && aadhaarField.style.display !== "none";
		const aadhaarNumber = aadhaarInput ? safeValue($(aadhaarInput).val()) : '';

		function focusFieldInTab(fieldSelector, tabIndex) {
		// Switch to the correct tab
		showTab(tabIndex);

		// Wait for tab content to render, then scroll and focus
		setTimeout(() => {
			const field = $(fieldSelector);
			if (field.length) {
				field.focus();
				$('html, body').animate({
					scrollTop: field.offset().top - 100
				}, 300);
			}
		}, 200);
	}

	const fieldTabMap = {
    // Personal Information tab (index 0)
    '#aadhaar_number_input': 0,
    '#current_city': 0,
	'#current_pin': 0,
	'#phone_number': 0,
	'#date_of_birth': 0,
	'#interviewed_date': 0,
	'#current_perm_post_office': 0,
	'#current_street': 0,
	'#current_house_no': 0,
	'#current_district': 0,
	'#current_state': 0,
	'#current_locality': 0,
    // Current Employer tab (index 1)
    '#manager_email': 1,
	'#manager_name': 1,
    '#manager_contact_no': 1,
    '#payslip_month_1': 1,

	};

		const date_of_birth = safeValue($('#date_of_birth').val());
			if (!date_of_birth) {
				showError('Please enter a valid date of birth.');
				focusFieldInTab('#date_of_birth', fieldTabMap['#date_of_birth']);
				return false;
			}

		if (isAadhaarVisible) {
			if (!aadhaarNumber || aadhaarNumber.length !== 12 || !/^\d{12}$/.test(aadhaarNumber)) {
				showError('Please enter a valid 12-digit Aadhaar number.');
				focusFieldInTab('#aadhaar_number_input', fieldTabMap['#aadhaar_number_input']);				return false;
			}
		}

		window.showTab = showTab;

		const phoneNumber = safeValue($('#phone_number').val());
		const phoneNumberRegex = /^\d{10}$/;
		const current_pin = safeValue($('#current_pin').val());
		const pinRegex = /^\d{6}$/;
		const current_house_no = safeValue($('#current_house_no').val());
		if (!current_house_no) {
			showError('Please enter a valid house number.');
			focusFieldInTab('#current_house_no', fieldTabMap['#current_house_no']);
			return false;
		}
		const current_city = safeValue($('#current_city').val());
		if (!current_city) {
			showError('Please enter a valid city.');
			focusFieldInTab('#current_city', fieldTabMap['#current_city']);
			return false;
		}
		const current_perm_post_office = safeValue($('#current_perm_post_office').val());
		if (!current_perm_post_office) {
			showError('Please enter a valid post office.');
			focusFieldInTab('#current_perm_post_office', fieldTabMap['#current_perm_post_office']);
			return false;
		}
		const current_street = safeValue($('#current_street').val());
		if (!current_street ) {
			showError('Please enter a valid street.');
			focusFieldInTab('#current_street', fieldTabMap['#current_street']);
			return false;
		}
		const current_district = safeValue($('#current_district').val());
		if (!current_district) {
			showError('Please enter a valid district.');
			focusFieldInTab('#current_district', fieldTabMap['#current_district']);
			return false;
		}

		if (!current_pin) {
			showError('Please enter a valid 6-digit PIN.');
			focusFieldInTab('#current_pin', fieldTabMap['#current_pin']);
			return false;
		}

		if (!pinRegex.test(current_pin)) {
			showError('Please enter a valid 6-digit PIN.');
			focusFieldInTab('#current_pin', fieldTabMap['#current_pin']);
			return false;
		}

		const current_locality = safeValue($('#current_locality').val());
		if (!current_locality) {
			showError('Please enter a valid locality.');
			focusFieldInTab('#current_locality', fieldTabMap['#current_locality']);
			return false;
		}

		const current_state = safeValue($('#current_state').val());
		if (!current_state) {
			showError('Please enter a valid state.');
			focusFieldInTab('#current_state', fieldTabMap['#current_state']);
			return false;
		}

		if (!phoneNumber || !phoneNumberRegex.test(phoneNumber)) {
			showError('Please enter a valid 10-digit mobile number.');
			focusFieldInTab('#phone_number', fieldTabMap['#phone_number']);
			return false;
		}

        const resultpercentage = safeValue($('#results_1').val());  // Sanitize input value
        // Check if the value is greater than 100
        if (resultpercentage && parseFloat(resultpercentage) > 100) {
            showError('Please enter a percentage less than or equal to 100.');  // Error message if value > 100
            focusFieldInTab('#results_1', fieldTabMap['#results_1']);  // Focus on the input field
            event.preventDefault();  // Prevent form submission
            return false;  // Stop further actions
        }

		// Only validate if currentEmployerTab is visible
		if (currentEmployerTab && currentEmployerTab.style.display !== "none") {
			const managerEmail = $('#manager_email').val();
			const managerName = safeValue($('#manager_name').val());
			const managerPhone = safeValue($('#manager_contact_no').val());
			const payslip_month_1= safeValue($('#payslip_month_1').val());
			const payslip_month_2= safeValue($('#payslip_month_2').val());
			const payslip_month_3= safeValue($('#payslip_month_3').val());

			const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
			const phoneRegex = /^\d{10}$/;

			if (!managerName) {
				showError('Manager Name is required.');
				focusFieldInTab('#manager_name', 1);
				return false;
			}

			if (!managerPhone) {
				showError(' Manager Contact Number is required.');
				focusFieldInTab('#manager_contact_no', 1);
				return false;
			}

			if (!phoneRegex.test(managerPhone)) {
				showError('Please enter a valid 10-digit mobile number for Manager Contact No.');
				focusFieldInTab('#manager_contact_no', 1);
				return false;
			}

			if (!managerEmail) {
				showError('Manager Email is required.');
				focusFieldInTab('#manager_email', 1);
				return false;
			}

			if (!emailRegex.test(managerEmail)) {
				showError('Please enter a valid manager email address.');
				focusFieldInTab('#manager_email', 1);
				return false;
			}

			if (!payslip_month_1 || !payslip_month_2 || !payslip_month_3) {
				showError('Please upload last 3 months payslips.');
				focusFieldInTab('#payslip_month_1', 1);
				return false;
			}
			const language = safeValue($('select[name="language"]').val()).trim();
			const speak = $('input[name="speak"]:checked').val() || '';
			const read = $('input[name="read"]:checked').val() || '';
			const write = $('input[name="write"]:checked').val() || '';

			//  Achievements (Professional / Awards)
			if ($('#achievements_checkbox').is(':checked')) {
				if (!safeValue($('#other_achievments').val().trim())) {
					showError('Please provide details of your achievements or awards.');
					focusFieldInTab('#other_achievments', 2);
					return false;
				}
			}

			//  Interviewed Before
			if ($('#interviewed_before_checkbox').is(':checked')) {
				const interviewFields = ['#position', '#interviewed_location', '#interviewed_date', '#interviewed_outcome'];
				for (let field of interviewFields) {
					if (!safeValue($(field).val().trim())) {
						showError('Please fill all interview details.');
						focusFieldInTab(field, 2);
						return false;
					}
				}
			}
			if (!language && (speak !== '' || read !== '' || write !== '')) {
				showError('Please select a language.');
				focusFieldInTab('select[name="language"]', 2);
				return false;
			}

			// --- Checkbox-driven mandatory validation ---

			//  Related to employee
			if ($('#related_to_employee_checkbox').is(':checked')) {
				const relatedFields = ['#related_employee', '#related_employee_org', '#related_employee_pos'];
				for (let field of relatedFields) {
					if (!safeValue($(field).val().trim())) {
						showError('Please fill all related employee details.');
						focusFieldInTab(field, 2); // Personal Info tab (adjust if needed)
						return false;
					}
				}
			}

			//  Professional Organization
			if ($('#professional_org_checkbox').is(':checked')) {
				if (!safeValue($('#professional_org').val().trim())) {
					showError('Please provide details of the Professional Organization.');
					focusFieldInTab('#professional_org', 2);
					return false;
				}
			}

			//  Political Organization
			if ($('#political_org_checkbox').is(':checked')) {
				if (!safeValue($('#political_org').val().trim())) {
					showError('Please provide details of the Political Organization.');
					focusFieldInTab('#political_org', 2);
					return false;
				}
			}

			//  Specialized Training
			if ($('#specialised_training_checkbox').is(':checked')) {
				if (!safeValue($('#specialised_training').val().trim())) {
					showError('Please provide details of the Specialized Training.');
					focusFieldInTab('#specialised_training', 2);
					return false;
				}
			}

		}
		event.preventDefault();
		const fields = [
			'father_name', 'applicant_name', 'date_of_birth', 'gender', 'country', 'marital_status',
			'current_house_no','current_city','current_perm_post_office','current_street','current_district',
			'current_pin','current_locality','current_state','period_years','current_period_months',
			'permanent_house_no','permanent_city','permanent_perm_post_office','permanent_street',
			'permanent_district','permanent_pin','permanent_locality','permanent_state', 'email_id',
			'aadhaar_number_input', 'name_of_employer','current_department','current_designation','reports_to',
			'manager_name','manager_contact_no','manager_email','reference_taken',
			'address_of_employer', 'duties_and_responsibilities', 'reason_for_leaving',
			'agency_details', 'current_salary', 'expected_salary',
			'other_achievments', 'position', 'interviewed_location', 'interviewed_date',
			'interviewed_outcome', 'related_employee', 'related_employee_org', 'related_employee_pos',
			'related_employee_rel', 'professional_org', 'political_org', 'specialised_training',
			'reference_taken', 'was_this_position', 'state_restriction', 'achievements_checkbox',
			'interviewed_before_checkbox', 'related_to_employee_checkbox', 'professional_org_checkbox',
			'political_org_checkbox', 'specialised_training_checkbox','additional_comments','phone_number'
		];

		const first_name = safeValue($('#first_name').val());
		const last_name = safeValue($('#last_name').val());
		const full_name = `${first_name} ${last_name}`.trim();
		const email = safeValue($('#email').val());

		const data = fields.reduce((obj, field) => {
			obj[field] = safeValue($(`#${field}`).val());
			return obj;
		}, {
			docname: applicant_id,
			date_of_birth: date_of_birth,
			interviewed_date: interviewed_date
		});

		data.in_india = $('#in_india_checkbox').is(':checked') ? 1 : 0;
		data.abroad = $('#abroad_checkbox').is(':checked') ? 1 : 0;
		data.is_form_submitted = $('#confirm').is(':checked') ? 1 : 0;

		data.education_qualification = [];
		$('#education_qualification_table tbody tr').each(function () {
			const fileInput = $(this).find('input[type="file"]')[0];
			const row = {
				course: safeValue($(this).find('.course').val()),
				name_of_school_college: safeValue($(this).find('.name_of_school_college').val()),
				name_of_universityboard_of_exam: safeValue($(this).find('.name_of_universityboard_of_exam').val()),
				dates_attended_from: safeValue($(this).find('.dates_attended_from').val()),
				dates_attended_to: safeValue($(this).find('.dates_attended_to').val()),
				result: safeValue($(this).find('.result').val()),
				attachments: fileInput ? fileInput.filedata : null
			};
			data.education_qualification.push(row);
		});

		data.professional_certification = [];
		$('#professional_qualification_table tbody tr').each(function () {
			const fileInput = $(this).find('input[type="file"]')[0];
			const row = {
				course: safeValue($(this).find('.course').val()),
				institute_name: safeValue($(this).find('.institute_name').val()),
				dates_attended_from: safeValue($(this).find('.dates_attended_from').val()),
				dates_attended_to: safeValue($(this).find('.dates_attended_to').val()),
				type_of_certification: safeValue($(this).find('.type_of_certification').val()),
				subject_major: safeValue($(this).find('.subject_major').val()),
				attachments: fileInput ? fileInput.filedata : null
			};
			data.professional_certification.push(row);
		});

		data.prev_emp_his = [];
		$('#previous_emplyoment_history_table tbody tr').each(function () {
			const fileInput = $(this).find('input[type="file"]')[0];
			const row = {
				name_of_org: safeValue($(this).find('.name_of_org').val()),
				prev_designation: safeValue($(this).find('.prev_designation').val()),
				last_salary_drawn: safeValue($(this).find('.last_salary_drawn').val()),
				name_of_manager: safeValue($(this).find('.name_of_manager').val()),
				period_of_employment: safeValue($(this).find('.period_of_employment').val()),
				reason_for_leaving: safeValue($(this).find('.reason_for_leaving').val()),
				attachments: fileInput ? fileInput.filedata : null
			};
			data.prev_emp_his.push(row);
		});

		data.language_proficiency = [];
		$('#table_3 tbody tr').each(function () {
			const row = {
				language: safeValue($(this).find('select[name="language"]').val()),
				speak: safeValue($(this).find('input[name^="speak"]:checked').val() || 0),
				read: safeValue($(this).find('input[name^="read"]:checked').val() || 0),
				write: safeValue($(this).find('input[name^="write"]:checked').val() || 0)
			};
			if (row.language) {
				data.language_proficiency.push(row);
			}
		});

		const payslipFields = ['payslip_month_1', 'payslip_month_2', 'payslip_month_3'];
		payslipFields.forEach(field => {
			const fileInput = $(`#${field}`)[0];
			if (fileInput && fileInput.files.length) {
				data[field] = fileInput.filedata;
			} else {
				data[field] = null;
			}
		});

		$('#form-error').hide();
		frappe.call({
			method: 'beams.www.job_application_upload.upload_doc.update_register_form',
			args: {
				form_data: JSON.stringify(data),
				docname: applicant_id
			},
			callback: function (r) {
				alert(r.message === 'success' ? 'Job Applicant updated successfully!' : 'Submission completed.');

				const confirmCheckbox = $('#confirm');
				const errorMessage = $('#error-message');

				if (confirmCheckbox.is(':checked')) {
					errorMessage.hide();
					window.location.href = '/application_success/success.html';
				} else {
					errorMessage.show(); // Show red message if checkbox not ticked
				}
			},
			error: function (err) {
				showError('An error occurred during submission.');
			}
		});
	});
});